serial_x.go

```diff
@@ -0,0 +1,116 @@
+// +build !windows
+
+package serial
+
+import (
+	"time"
+	"io/ioutil"
+	"regexp"
+	"strings"
+)
+
+const (
+	devFolder = "/dev"
+	regexFilter = `^(cu|tty)\.?.*`
+)
+
+/*
+This function was taken with minor modifications from the go.bug.st/serial package (https://github.com/bugst/go-serial), and is subject to the conditions of its license (reproduced below):
+
+Copyright (c) 2014, Cristian Maglie.
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+   notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+   notice, this list of conditions and the following disclaimer in
+   the documentation and/or other materials provided with the
+   distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+   contributors may be used to endorse or promote products derived
+   from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+*/
+func GetPortsList() ([]string, error) {
+	files, err := ioutil.ReadDir(devFolder)
+	if err != nil {
+		return nil, err
+	}
+
+	ports := make([]string, 0, len(files))
+	for _, f := range files {
+		// Skip folders
+		if f.IsDir() {
+			continue
+		}
+
+		// Keep only devices with the correct name
+		match, err := regexp.MatchString(regexFilter, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if !match {
+			continue
+		}
+
+		portName := devFolder + "/" + f.Name()
+
+		// Check if serial port is real or is a placeholder serial port "ttySxx"
+		if strings.HasPrefix(f.Name(), "ttyS") {
+			port, err := openPort(portName, 9600, 100)
```

Comment:
It's exactly that -- I've been developing embedded firmware for a project involving a couple of microcontrollers, one of which features a bootloader and some configuration via a USB terminal connection. Hacked together a little GUI tool as a learning exercise in Go which acts as a configuration frontend and firmware upgrade tool. It's internal-use only, hence the poor quality control :)  Users are myself (running linux), the hardware developer and one other non-technical colleague (both of whom run windows). Automatic identification of the serial port (via some additional regex filtering outside of the serial library, an attempted connection and ping) simplified use for the non-technical colleague.

Didn't think it would be of any wider interest until I noticed the issue the other day, and may well not be.